### ISSUE TITLE
BUG: dataframe concatenation segementation fault (core dump)

### ISSUE THREAD

**Issue by stokhos** [pandas-dev/pandas#39144]
- [ X ] I have checked that this issue has not already been reported.

- [ X ] I have confirmed this bug exists on the latest version of pandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

**Note**: Please read [this guide](https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports) detailing how to provide the necessary information for us to reproduce your bug.

#### Code Sample, a copy-pastable example

```python
    df = pd.concat(
        (
            pd.read_csv(os.path.join(dirpath, filename))
            for dirpath, filename in file_generator
        ),
        ignore_index=True,
    )


```

#### Problem description

A subgroup of my files can cause the above function to break. And it happens 100% with those files with pandas1.2.0, with 1.1.5 no problem.

[this should explain **why** the current behaviour is a problem and why the expected output is a better solution]

#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here leaving a blank line after the details tag]

</details>
INSTALLED VERSIONS
------------------
commit           : 3e89b4c4b1580aa890023fc550774e63d499da25
python           : 3.9.1.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.9.16-200.fc33.x86_64
Version          : #1 SMP Mon Dec 21 14:08:22 UTC 2020
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.2.0
numpy            : 1.19.5
pytz             : 2020.5
dateutil         : 2.8.1
pip              : 20.2.2
setuptools       : 49.1.3
Cython           : 0.29.21
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.5.1
html5lib         : 1.1
pymysql          : None
psycopg2         : None
jinja2           : 2.11.2
IPython          : 7.18.1
pandas_datareader: 0.9.0
bs4              : 4.9.3
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.3.3
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.5.2
sqlalchemy       : None
tables           : None
tabulate         : 0.8.7
xarray           : None
xlrd             : None
xlwt             : None
numba            : None

**Comment by phofl** [comment]
Please provide a reproducible example https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

**Comment by stokhos** [comment]
@phofl  Is there a way for me to pin down the bug on my side? It only happens to to files in two folder. I can't do the minimal bug report, because it has more than 1000 files.

**Comment by phofl** [comment]
You could do a git bisect to narrow down when this was introduced

IAlternatively it should be sufficientto use only these files? Maybe you could create the DataFrames casing the segfaults in code?

**Comment by jorisvandenbossche** [comment]
@stokhos a good start would probably to check whether the segfault happens in the concat function or actually in the `read_csv` (my guess would be the second). If you rewrite your code a little bit in multiple lines, you should be able to see on which line it fails. 

In case it's caused by `read_csv`, you could then try to find out which of the 1000 files is failing to read (eg read them in a for loop and print the file path before reading, to see which one is failing)

**Comment by MarcoGorelli** [comment]
Closing for now, but please do let us know if you have a reproducible example, as per:

> Please provide a reproducible example https://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports

**Comment by djherbis** [comment]
I was able to reliably reproduce a segfault with a single row CSV on read_csv on pandas 1.2.0, however upgrading to 1.2.3 fixed it.

**Comment by MarcoGorelli** [comment]
@djherbis could you share that csv please?

**Comment by djherbis** [comment]
It's unfortunately got some private data in it, I'll try to narrow it down to single row, single column, and scrub the data so that it repros without private info.

**Comment by djherbis** [comment]
Looks like this may have been the same issue as https://github.com/pandas-dev/pandas/issues/38753#issuecomment-751985476 which was fixed since using the float_precision="legacy" worked even on 1.2.0.

**Comment by djherbis** [comment]
Yeah I think it's that issue, narrowed it down to just this:

```bash
echo "id" > crash.csv
echo "0E59181149492B9CCD27179CDB27EBEC" >> crash.csv
```

```python
import pandas as pd
pd.read_csv('crash.csv')
```

Looks like pandas treats that string as an exponent number.

**Comment by MarcoGorelli** [comment]
Thanks @djherbis - I'll do a bisect later, we may need to add a test for this

**Comment by djherbis** [comment]
You can check the other bug I linked too, I believe they may have already reported and fixed this particular issue.

**Comment by MarcoGorelli** [comment]
This was fixed in #38789
```
b438f747d436f7f25ba3e2e04ecf5d70c4121ab4 is the first bad commit
commit b438f747d436f7f25ba3e2e04ecf5d70c4121ab4
Author: mzeitlin11 <37011898+mzeitlin11@users.noreply.github.com>
Date:   Wed Dec 30 13:40:59 2020 -0500

    BUG: Fix precise_xstrtod segfault on long exponent (#38789)
```
so I think we can close - thanks anyway for the useful reproducer!
